aktool/aktool.1
'\" t
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "AKTOOL" "1" "нояб. 15, 2024" "версия библиотеки 0.9.17" "libakrypt"
.SH NAME
aktool \- Правила пользования
.sp
\fIaktool\fP \- консольная утилита, предназначенная для криптографической защиты данных в пространстве пользователя.
.sp
Утилита предоставляет интерфейс к различным криптографическим
преобразованиям таким, как шифрование и контроль целостности данных,
выработка и проверка электронной подписи, генерация криптографических
ключей и паролей, создание и верификация сертификатов открытых ключей.
.SH ОПИСАНИЕ
.sp
Синтаксис выполнения программы:
.INDENT 0.0
.INDENT 3.5
.sp
.EX
aktool команда [опции] [файл или каталог]
.EE
.UNINDENT
.UNINDENT
.sp
Каждая команда определяет отдельный класс криптографических
преобразований и функциональных возможностей. В настоящее время
доступны следующие команды.
.TS
box center;
l|l|l.
T{
Краткая форма
T}	T{
Полная форма
T}	T{
Описание
T}
_
T{
a
T}	T{
asn1parse
T}	T{
Декодирование и печать ASN.1 данных
T}
_
T{
e
T}	T{
encrypt
T}	T{
Зашифрование данных пользователя
T}
_
T{
d
T}	T{
decrypt
T}	T{
Расшифрование данных пользователя
T}
_
T{
i
T}	T{
icode
T}	T{
Вычисление контрольной суммы или имитовставки
T}
_
T{
k
T}	T{
key
T}	T{
Управление ключевой информацией
T}
_
T{
s
T}	T{
show
T}	T{
Вывод служебной и справочной информации
T}
_
T{
t
T}	T{
test
T}	T{
Тестирование корректности и скорости работы
T}
.TE
.sp
Получить информацию о доступных опциях и параметрах любой
команды можно с помощью вызова:
.INDENT 0.0
.INDENT 3.5
.sp
.EX
aktool команда \-\-help
.EE
.UNINDENT
.UNINDENT
.sp
Получить информацию о текущей версии утилиты и компиляторе, которым она была собрана,
можно с помощью вызова:
.INDENT 0.0
.INDENT 3.5
.sp
.EX
aktool \-\-version
.EE
.UNINDENT
.UNINDENT
.SH УПРАВЛЕНИЕ КЛЮЧЕВОЙ ИНФОРМАЦИЕЙ
.sp
Команда \fIkey\fP (короткая форма \fIk\fP) позволяет вырабатывать секретные и открытые ключи
для дальнейшего использования в алгоритмах шифрования, контроля целостности,
процедурах выработки и проверки электронной подписи и т.п.
.sp
Поддерживается инфраструктура сертификатов открытых ключей, а также
генерация ключей парной связи (симметричных ключей)
в соответствиии со схемой Блома распределения ключевой информации.
.sp
Команда \fIkey\fP позволяет сохранять секретные ключи в файлы или
ключевое хранилище в контейнере, формат которого является открытым
и распространяется вместе с документацией к библиотеке \fIlibakrypt\fP\&.
Перед сохранением секретного ключа пользователю предлагается
ввести пароль, используемый для шифрования созданного секретного ключа.
Ввод пароля можно осуществлять как в символьном, так и в шестнацатеричном формате.
.sp
Вырабатываемые открытые ключи могут сохраняться как в формате запроса на сертификат,
так и в формате собственно сертификата в соответствии с рекомендациями
Р 1323565.1.23\-2018. Также команда \fIkey\fP позволяет просматривать и
изменять параметры созданных ранее ключей.
.SS Опции команды key
.INDENT 0.0
.TP
.B \-a, \-\-algorithm=метод
.UNINDENT
.sp
Опция позволяет определить метод или алгоритм, используемый для генерации ключевой
информации. Для этого необходимо указать его имя или идентификатор.
Опция необходима в некоторых схемах выработки ключей, например, при использовании \fI\%схемы Блома\fP\&.
.INDENT 0.0
.TP
.B \-\-ca\-cert=файл
.UNINDENT
.sp
Опция позволяет указать имя файла, содержащего сертификат открытого ключа центра сертификации.
Опция может использоваться для выработки и верификации пользовательских сертификатов.
.INDENT 0.0
.TP
.B \-\-ca\-key=файл
.UNINDENT
.sp
Опция позволяет указать имя файла, содержащего секретный ключ центра сертификации.
Опция может применяться при генерации пользовательских сертификатов из запросов на сертификат.
Данная опция является другой формой опции \fI\%\-\-key\fP\&.
.INDENT 0.0
.TP
.B \-\-curve=имя
.UNINDENT
.sp
Опция позволяет указать имя или идентификатор эллиптической кривой,
которой принадлежит открытый ключ.
.INDENT 0.0
.TP
.B \-\-days=число
.UNINDENT
.sp
Опция позволяет определить срок действия секретного или открытого ключа.
Срок исчисляется заданным количеством дней, начиная с момента вызова утилиты.
.INDENT 0.0
.TP
.B \-\-field=[ 256, 512 ]
.UNINDENT
.sp
Опция позволяет указать \fIn\fP \- степень расширения конечного поля \emathbb F_{2^n}, используемого для реализации
криптографических алгоритмов. Допустимыми аргументами опции являются значения 256 или 512 бит.
.INDENT 0.0
.TP
.B \-\-format=[ der, pem, certificate ]
.UNINDENT
.sp
Опция позволяет указать формат хранения вырабатываемого файла с ключевой информацией.
Допустимыми аргументами опции являются значения \fIder\fP, \fIpem\fP или \fIcertificate\fP\&.
.INDENT 0.0
.TP
.B \-\-id=строка
.UNINDENT
.sp
Опция позволяет определить имя или идентификатор владельца ключа.
.sp
В случае использования данной опции одовременно с созданием сертификата открытого ключа или запроса на сертификат
значение \fIстрока\fP определяет расширенное имя владельца.
Расширенное имя владельца должно состоять из последовательности секций вида
\fI/секция1=значение/секция2=значение2\fP,
где название секции должно принимать значение из множества:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fIem\fP (электронная почта), \fIcn\fP (основное имя), \fIct\fP (страна), \fIst\fP (область), \fIlt\fP (поселение), \fIsa\fP (улица), \fIor\fP (организация), \fIou\fP (отдел), \fItl\fP (должность), \fIsu\fP (фамилия), \fIgn\fP (имя), \fIps\fP (псевдоним) и \fIsn\fP (серийный номер).
.UNINDENT
.UNINDENT
.UNINDENT
.sp
Также поддерживаются названия секций, рекомендованные для использования в Российской Федерации:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fIog\fP (ОГРН), \fIoi\fP (ОГРНИП), \fIsi\fP (СНИЛС), \fIin\fP (ИНН физического лица) и \fIle\fP (ИНН юридического лица).
.UNINDENT
.UNINDENT
.UNINDENT
.sp
В случае создания сертификатов или запросов на сертификат открытого ключа
допускается использование опции  без указания имени секции. В этом случае
переданная строка определяет основное имя владельца ключа.
.INDENT 0.0
.TP
.B \-\-id\-hex=строка
.UNINDENT
.sp
Опция позволяет указать имя владельца секретного ключа в виде шестнадцатеричной последовательности.
Данная опция не применяется при выработке сертификатов или запросов на сертификат открытого ключа.
.INDENT 0.0
.TP
.B \-\-inpass=строка
.UNINDENT
.sp
Опция позволяет указать пароль доступа к считываемому секретному ключу в командной строке.
Использование данной опции может привести к компрометации ключевой информации.
.INDENT 0.0
.TP
.B \-\-inpass\-hex=строка
.UNINDENT
.sp
Опция позволяет указать пароль доступа к считываемому секретному ключу в командной строке
в виде шестнадцатеричной последовательности.
Использование данной опции может привести к компрометации ключевой информации.
.INDENT 0.0
.TP
.B \-\-key=файл
.UNINDENT
.sp
Опция позволяет указать имя файла, содержащего считываемый секретный ключ
симметричного или асимметричного алгоритма.
.INDENT 0.0
.TP
.B \-\-label=строка
.UNINDENT
.sp
Опция позволяет присвоить секретному ключу пользовательскую метку или описание.
.INDENT 0.0
.TP
.B \-n, \-\-new
.UNINDENT
.sp
Опция создает новый секретный ключ или ключевую пару для заданного криптографичеспкого алгоритма.
.INDENT 0.0
.TP
.B \-\-no\-outpass
.UNINDENT
.sp
Опция позволяет не указывать пароль доступа к контейнеру создаваемого секретного ключа.
В этом случае ключ помещается в контейнер в незашифрованном виде, что является потенциально небезопасным.
.INDENT 0.0
.TP
.B \-\-op
.UNINDENT
.sp
Короткая форма опции \fI\%\-\-output\-public\-key\fP\&.
.INDENT 0.0
.TP
.B \-\-outpass=строка
.UNINDENT
.sp
Опция позволяет указать пароль доступа к созданному секретному ключу прямо в командной строке.
Использование данной опции не является безопасным и может привести к компрометации ключевой ифнормации.
.INDENT 0.0
.TP
.B \-\-outpass\-hex=строка
.UNINDENT
.sp
Опция позволяет указать пароль доступа к созданному секретному ключу прямо в командной строке
в виде шестнадцатеричной последовательности.
Использование данной опции не является безопасным и может привести к компрометации ключевой ифнормации.
.INDENT 0.0
.TP
.B \-\-output\-public\-key=файл
.UNINDENT
.sp
Опция позволяет указать имя файла, в котором будет сохранен созданный сертификат или
запрос на сертификат открытого ключа. Имеет короткую форму \fI\%\-\-op\fP\&.
.INDENT 0.0
.TP
.B \-o, \-\-output\-secret\-key=файл
.UNINDENT
.sp
Опция позволяет указать имя файла, в котором будет сохранен созданный секретный ключ.
.INDENT 0.0
.TP
.B \-\-p7b\-create
.UNINDENT
.sp
Опция позволяет поместить один или несколько сертификатов центра сертификации в один контейнер
в формате PKCS #7 (обычно используется расширение \fIp7b\fP).
Помещаемые в контенер сертифкаты указываются в качестве последних аргументов командной строки.
.INDENT 0.0
.TP
.B \-\-p7b\-ls=файл
.UNINDENT
.sp
Опция позволяет вывести в консоль список сертификатов,
хранящихся в PKCS #7 контейнере.
.INDENT 0.0
.TP
.B \-\-p7b\-split=файл
.UNINDENT
.sp
Опция позволяет извлечь из контейнера в формате PKCS #7
все доступные сертификаты и сохранить их в виде отдельных файлов (в кодировке \fIder\fP или \fIpem\fP).
.INDENT 0.0
.TP
.B \-\-random=имя
.UNINDENT
.sp
Опция позволяет указать имя или идентификатор генератора псевдослучайных последовательностей,
используемого для выработки ключевой информации или электронной подписи.
По умолчанию устанавливается генератор, зависящий от операционной системы, в которой выполняется утилита.
В \fIUnix\fP\-системах это генератор, доступ к которому обеспечивается символьным устройством \fI/dev/random\fP\&.
.INDENT 0.0
.TP
.B \-\-random\-file=файл
.UNINDENT
.sp
Опция позволяет в явном виде указать имя файла из которого происходит чтение псевдослучайных последовательностей.
Использование данной опции целесообразно для доступа к физическим генераторам, предоставляющим доступ к вырабатываемым
данным посредством символьных устройств.
.INDENT 0.0
.TP
.B \-\-repo=каталог
.UNINDENT
.sp
Опция указывает путь к хранилищу доверенных сертификатов.

Changelog.md
#Libakrypt: перечень изменений


## Изменения в версии 0.9.17

 - добавлены опции cmake
   - AK_TOOL для включения/отключения сборки консольной утилиты
   - AK_BASE для сборки библиотеки в виде одного или двух исполняемых модулей
 - изменения в утилите aktool icode
   - добавлена опция -n, --no-database
   - добавлены опции --offset, --size для контроля целостности фрагментов файлов
 - исправлены ошибки, возникающие при компиляции библиотеки с флагами -Werror -Wextra -Wpedantic  


## Изменения в версии 0.9.16

 - исправлены некоторые ошибки при контроле целостности процессов в памяти
   для 32х битных архитектур
 - добавлены функции 
    - ak_file_lseek(), 
    - ak_mac_file_offset(), 
    - ak_hash_file_offset(), 
    - ak_hmac_file_offset() 
    - ak_bckey_cmac_file_offset()
    - ak_signkey_sign_file_offset()
    - ak_verifykey_verify_file_offset()
 - исправлена скрытая ошибка в функции ak_hash_context_streebog_sadd() для 
   поддержки некоторых 32х битных платформ 
 - изменения в утилите aktool icode
    - опция -o, --output заменена на опцию -d, --database
    - добавлена опция --exclude-link 
 - полностью удалена реализация методов асимметричного шифрования,
   причины:
   - невостребованость в текущем виде,
   - планируемое, в будущем, изменение реализации + модификация используемой схемы шифрования


## Изменения в версии 0.9.15

 - В библиотеку libakrypt-base добавлены функции 
    - ak_time_to_str()
    - ak_list_add_node_sorted()
 - Изменен интерфейс, а также реализованы дополнительные 
   функции выработки производных ключей (kdf256, kdf512 и kdfnmac):
    - ak_skey_derive_kdf_hmac()
    - ak_skey_derive_kdf_hmac_from_skey()
    - ak_skey_set_derive_kdf_hmac_from_skey()
 - Добавлена функция проверки статистической гипотезы о равномерном 
   распределении элементов двоичной последовательности (используется при динамическом анализе ПДСЧ)
    - ak_random_dynamic_test()
 - Добавлена поддержка новой опции библиотеки, отвечающей за дополнительные проверки
   корректной работы ряда криптографических алгоритмов


## Изменения в версии 0.9.14

 - Добавлена реализация хэш-таблицы с возможностью поиска данных,
   сохранения/чтения таблицы, а также проверки уникальности ключа поиска
 - Переписана большая часть команды icode для утилиты aktool
    - теперь поддерживается контроль целостности как загружаемых,
      так и уже загруженных сегментов исполняемого кода и данных,
    - реализована поддержка файлов конфигурации
 - Исправлена реализации функции ak_file_read_by_lines
 - Исправлена ошибка при чтении секретных ключей с паролем нулевой длины
 - Исправлена ошибка при отображении файлов в память (функция ak_file_mmap)
 - Уточнены имена ряда статических (неэкспортируемых) функций библиотеки
 - Исправлены неточности в механизме сборки документации


## Изменения в версии 0.9.13
 
 - Добавлены функции для чтения и записи открытых ключей в простом текстовом формате
     - ak_verifykey_create_from_file()
     - ak_verifykey_export_to_file()
 - Реализация некриптографической функции crc64 (стандарт: ECMA 182, Прил. B)
    - ak_hash_create_crc64()  
 - Добавлены функции, связывающие численные идентификаторы кривых и их параметры
    - ak_wcurve_get_identifier()
    - ak_wcurve_get_parameters()
 - Добавленный функционал утилиты aktool:
    - появилась возможность удаления файлов после изменения формата
      представления asn1 данных (pem или der)


## Изменения в версии 0.9.12

 - Преобразован раздел с функциями генерации производных ключей.
   Изменены или добавлены следующие функции:
    - int ak_skey_derive_kdf256()
    - int ak_skey_derive_kdf256_from_skey()
    - ak_pointer ak_skey_new_derive_kdf256_from_skey()
    - int ak_tlstree_state_create()
    - int ak_tlstree_state_destroy()
    - int ak_skey_derive_tlstree()
    - int ak_skey_derive_tlstree_from_skey()
    - ak_pointer ak_skey_new_derive_tlstree_from_skey()       
 - Встроены в общее дерево тестов криптографических преобразований:
    - ak_libakrypt_test_kdf256()
    - bool_t ak_libakrypt_test_tlstree()
 - Добавлен ряд забытых когда-то функций для кодирования/декодирования base64:
    - const char *ak_ptr_to_base64()
    - char *ak_ptr_to_base64_alloc()
    - size_t ak_ptr_to_base64_size()
    - ak_uint8 *ak_base64_to_ptr()   
 - Название функции ak_skey_delete_after_load() укоротилось до ak_skey_delete()
 - Изменена строка с авторами библиотеки в файле с лицензией


## Изменения в версии 0.9.11

 - Возвращена реализация ПДСЧ на основе функций хеширования (hrng)
 - Изменена команда для подписи сертификатов открытого клчюа в утилите aktool 
   (произведена замена --cert на --sign)
 - Добавлен ряд функций для работы с сертификатами, в том числе:
    - int ak_certificate_export_to_repository() 
    - int ak_certificate_add_file_to_repository()
    - int ak_certificate_add_ptr_to_repository()
    - int ak_certificate_add_asn1_to_repository()
    - int ak_certificate_import_from_repository
 - Добавлена возможность сборки библиотеки с musl вместо libc
 - Реализована поддержка petalinux
 - Существенно расширена документация к библиотеке
 
 
## Изменения в версии 0.9.10

 - Исправлена ошибка в реализации алгоритма nmac [функции finalize и create_nmac_...()]

aktool/aktool.c
/* ----------------------------------------------------------------------------------------------- */
/*  Copyright (c) 2014 - 2021 by Axel Kenzo, axelkenzo@mail.ru                                     */
/*                                                                                                 */
/*  Файл akrypt.c                                                                                  */
/*  - содержит реализацию консольной утилиты, иллюстрирующей возможности библиотеки libakrypt      */
/* ----------------------------------------------------------------------------------------------- */
 #include <aktool.h>
#ifdef _MSC_VER
 #include <strsafe.h>
#endif

/* ----------------------------------------------------------------------------------------------- */
 aktool_ki_t ki = {
   .aktool_log_level = ak_log_none, /* аудит не выполняется */
   .aktool_openssl_compability = ak_false,    /* совместимость с openssl не предусматривается */
   .aktool_hex_password_input = ak_false,    /* пароли вводятся как символы текста */
   .verbose = ak_false,    /* вывод подробной информации/справки  */
   .quiet = ak_false,
   .show_caption = ak_true,
   .method = NULL,
   .oid_of_generator = NULL,
   .name_of_file_for_generator = NULL,
   .generator = NULL,
   .oid_of_target = NULL,
   .format = asn1_der_format,
   .target_undefined = ak_false,
   .curve = NULL,
   .days = 365, /* срок действия ключа (в сутках) */
   .field = ak_galois256_size,
   .size = 512,
   .keylabel = NULL, /* указатель на внешнюю область памяти */
   .lenuser = 0,
   .leninpass = 0,
   .lenoutpass = 0,
   .userid = "",
   .inpass = "",
   .outpass = "",
   .audit_filename = "", /* имя файла аудита не определено */
   .os_file = "", /* выход, открытый ключ */
   .op_file = "", /* выход, секретный ключ */
   .key_file = "", /* вход, секретный ключ */
   .pubkey_file = "", /* вход, открытый ключ */
   .capubkey_file = "" /* вход, открытый ключ УЦ */
 };

/* ----------------------------------------------------------------------------------------------- */
/* функция, реализующая аудит */
 ak_function_log *audit =
#ifdef _WIN32
 aktool_audit_function;
#else
 ak_function_log_syslog;
#endif

#ifdef _WIN32
 unsigned int aktool_console_page = 0;
#endif

/* ----------------------------------------------------------------------------------------------- */
 int main( int argc, tchar *argv[] )
{
 /* определение переменных, используемых для указания времени старта программы */
 #ifdef _WIN32
  time_t ptime;
  TCHAR homepath[FILENAME_MAX]
   #ifdef _MSC_VER
     , buffer[64]
   #endif
   ;
 #endif
 /* попытка русификации программы для unix-like операционных систем */
 #ifdef AK_HAVE_LIBINTL_H
 /* обрабатываем настройки локали
    при инсталляции файл aktool.mo должен помещаться в /usr/share/locale/ru/LC_MESSAGES
    (однако, при использовании bsd-like адрес может быть другим) */
  #ifdef AK_HAVE_LOCALE_H
   setlocale( LC_ALL, "" );
  #endif
  bindtextdomain( "aktool", LIBAKRYPT_LOCALE_PATH ); /* вместо фиксированного /usr/share/locale */
  textdomain( "aktool" );
 #endif

 /* проверяем, что пользователем должна быть задана команда */
  if( argc < 2 ) return aktool_litehelp();

 /* проверяем флаги вывода справочной информации */
  if( aktool_check_command( "-h", argv[1] )) return aktool_help();
  if( aktool_check_command( "--help", argv[1] )) return aktool_help();
  if( aktool_check_command( "/?", argv[1] )) return aktool_help();
  if( aktool_check_command( "--version", argv[1] )) return aktool_version();

 #ifdef _WIN32
  if( ak_homepath( homepath, FILENAME_MAX ) == ak_error_ok ) {
    ak_snprintf( ki.audit_filename, sizeof( ki.audit_filename ),
                                                  "%s\\.config\\libakrypt\\aktool.log", homepath );
    remove( ki.audit_filename );
    ak_log_set_function( audit = aktool_audit_function );

   #ifdef _MSC_VER
    _time64( &ptime );
    _tctime64_s( buffer, sizeof( buffer ), &ptime );
    ak_snprintf( homepath, FILENAME_MAX, "%s started at %s", argv[0], buffer );
   #else
     ptime = time( NULL );
     ak_snprintf( homepath, FILENAME_MAX, "%s started at %s", argv[0], ctime( &ptime ));
   #endif
    ak_log_set_message( homepath );
  }
 #endif

 /* выполняем команду пользователя */
  if( aktool_check_command( "s", argv[1] )) return aktool_show( argc, argv );
  if( aktool_check_command( "show", argv[1] )) return aktool_show( argc, argv );
  if( aktool_check_command( "a", argv[1] )) return aktool_asn1( argc, argv );
  if( aktool_check_command( "asn1parse", argv[1] )) return aktool_asn1( argc, argv );
  if( aktool_check_command( "t", argv[1] )) return aktool_test( argc, argv );
  if( aktool_check_command( "test", argv[1] )) return aktool_test( argc, argv );
  if( aktool_check_command( "k", argv[1] )) return aktool_key( argc, argv );
  if( aktool_check_command( "key", argv[1] )) return aktool_key( argc, argv );
  if( aktool_check_command( "i", argv[1] )) return aktool_icode( argc, argv );
  if( aktool_check_command( "icode", argv[1] )) return aktool_icode( argc, argv );

 /* ничего не подошло, выводим сообщение об ошибке */
  ak_log_set_function( ak_function_log_stderr );
  ak_error_message_fmt( ak_error_undefined_function,
                                                 __func__, _("undefined command \"%s\""), argv[1] );
 return EXIT_FAILURE;
}

/* ----------------------------------------------------------------------------------------------- */
/*                         реализация функций обработки команд пользователя                        */
/* ----------------------------------------------------------------------------------------------- */
 bool_t aktool_check_command( const char *comm, tchar *argv )
{
 size_t len = strlen( comm );

  if( strlen( argv ) != len ) return ak_false;
  if( strncmp( comm, argv, len )) return ak_false;
 return ak_true;
}

/* ----------------------------------------------------------------------------------------------- */
/* вывод сообщений в заданный пользователем файл, а также в стандартный демон вывода сообщений */
 int aktool_audit_function( const char *message )
{
  FILE *fp = fopen( ki.audit_filename, "a+" );
   /* функция выводит сообщения в заданный файл */
    if( !fp ) return ak_error_open_file;
    fprintf( fp, "%s\n", message );
    if( fclose(fp) == EOF ) return ak_error_access_file;

 return ak_error_ok;
}

/* ----------------------------------------------------------------------------------------------- */
 void aktool_set_audit( tchar *message )
{
  if( ak_ptr_is_equal( "stderr", message, 6 )) {
       audit = ak_function_log_stderr; /* если задан stderr, то используем готовую функцию */
  } else {
            if( strlen( message ) > 0 ) {
              memset( ki.audit_filename, 0, sizeof( ki.audit_filename ));
              strncpy( ki.audit_filename, message, sizeof( ki.audit_filename )-1 );
              audit = aktool_audit_function;
            }
         }
}

/* ----------------------------------------------------------------------------------------------- */
 void aktool_error( const char *format, ... )
{
  va_list args;
  int result = 0;
  char string[1024];

  va_start( args, format );
 #ifdef _MSC_VER
  #if _MSC_VER > 1310
    result = _vsnprintf_s( string, sizeof( string ), sizeof( string ), format, args );
  #else
    result = _vsnprintf( string, sizeof( string ), format, args );
  #endif
 #else
  result = vsnprintf( string, sizeof( string ), format, args );
 #endif
  va_end( args );

 if( !ki.quiet ) {
   if( result >= 0 ) /* выводим сообщение только, если нет режима тишины */
     printf(_("%serror%s: %s\n"), ak_error_get_start_string(), ak_error_get_end_string(), string );
 }
}

/* ----------------------------------------------------------------------------------------------- */
 bool_t aktool_create_libakrypt( void )
{
  ak_int64 number;

 /* устанавливаем уровень аудита */
  ak_log_set_level( ki.aktool_log_level );

 /* инициализируем библиотеку */
  if( ak_libakrypt_create( audit ) != ak_true ) {
    ak_libakrypt_destroy();
    aktool_error(_("incorrect initialization of libakrypt library"));
    return ak_false;
  }

 /* применяем флаг совместимости с openssl */
  number = ak_libakrypt_get_option_by_name( "openssl_compability ");
  if(( number != ak_error_wrong_option ) && ( ki.aktool_openssl_compability != number ))
    ak_libakrypt_set_openssl_compability( ki.aktool_openssl_compability );

#ifdef _WIN32
  aktool_console_page = GetConsoleCP();
  /*SetConsoleCP( 65001 ); SetConsoleOutputCP( 65001 ); */
  SetConsoleCP( 1251 ); SetConsoleOutputCP( 1251 );
#endif

 return ak_true;
}

/* ----------------------------------------------------------------------------------------------- */
 int aktool_destroy_libakrypt( void )
{
 #ifdef _WIN32
  SetConsoleCP( aktool_console_page );
  SetConsoleOutputCP( aktool_console_page );
 #endif
 return ak_libakrypt_destroy();
}


/* ----------------------------------------------------------------------------------------------- */
 int aktool_print_gettext( const char *message )
{
  if( message ) fprintf( stdout, "%s", _( message ));
 return ak_error_ok;
}

/* ----------------------------------------------------------------------------------------------- */
/*                            функции генерации ключевой информации                                */
/* ----------------------------------------------------------------------------------------------- */
 ak_random aktool_key_new_generator( void )
{
  ak_random generator = NULL;

  if( ki.name_of_file_for_generator != NULL ) {
    if( ak_random_create_file( generator = malloc( sizeof( struct random )),
                                                ki.name_of_file_for_generator ) != ak_error_ok ) {
      if( generator ) free( generator );
      return NULL;
    }
    if( ki.verbose ) printf(_("using file with random data: %s\n"),
                                                                  ki.name_of_file_for_generator );
  }
   else {
    if(( generator = ak_oid_new_object( ki.oid_of_generator )) == NULL ) return NULL;
    if( ki.verbose ) printf(_("using random number generator: %s\n"),
                                                                    ki.oid_of_generator->name[0] );
   }

 return generator;
}

/* ----------------------------------------------------------------------------------------------- */
 void aktool_key_delete_generator( ak_random generator )
{
   if( ki.name_of_file_for_generator != NULL ) {
     ak_random_destroy( generator );
     free( generator );
   }
    else ak_oid_delete_object( ki.oid_of_generator, generator );
}

/* ----------------------------------------------------------------------------------------------- */
/*                                    Функции ввода ключей                                         */
/* ----------------------------------------------------------------------------------------------- */
/*! Эта функция используется перед импортом ключа */
 ssize_t aktool_load_user_password( const char *pts, char *password,
                                                        const size_t pass_size , password_t type )
{
  (void)pts;
  (void)type;

 /* копируем пароль, если он уже был установлен */
  memset( password, 0, pass_size );
  if( ki.leninpass > 0 ) {
    memcpy( password, ki.inpass, ak_min( (size_t)ki.leninpass, pass_size ));
    return ki.leninpass;
  }

  if( ki.aktool_hex_password_input ) {
    return ak_password_read_from_terminal(_("Input password [as hexademal string]: "),
                                                             password, pass_size, hexademal_pass );
  }
 return ak_password_read_from_terminal(_("Input password: "), password, pass_size, symbolic_pass );
}

/* ----------------------------------------------------------------------------------------------- */
/*! Эта функция используется перед экспортом ключа */
 ssize_t aktool_load_user_password_twice( char *password, const size_t pass_size )
{
  ssize_t buflen1, buflen2;
  char buf1[aktool_password_max_length], buf2[aktool_password_max_length];

 /* считываем первый раз */
  if( ki.aktool_hex_password_input ) {
    buflen1 = ak_password_read_from_terminal(_("Input password [as hexademal string]: "),
                                                              buf1, sizeof(buf1), hexademal_pass );
  }
   else buflen1 = ak_password_read_from_terminal(_("Input password: "),
                                                               buf1, sizeof(buf1), symbolic_pass );
  if( buflen1 < 1 ) {
    aktool_error(_("password has zero length"));
    return buflen1;
  }

 /* считываем второй раз */
  if( ki.aktool_hex_password_input ) {
    buflen2 = ak_password_read_from_terminal(_("Retype password [as hexademal string]: "),
                                                              buf2, sizeof(buf2), hexademal_pass );
  }
   else buflen2 = ak_password_read_from_terminal(_("Retype password: "),
                                                               buf2, sizeof(buf2), symbolic_pass );

  if(( buflen1 != buflen2 ) || ( !ak_ptr_is_equal( buf1, buf2, buflen1 ))) {
      aktool_error(_("the passwords don't match"));
      buflen1 = ak_error_not_equal_data;
  }
   else {
           memset( password, 0, pass_size );
           memcpy( password, buf1, buflen1 = ak_min(( size_t )buflen1, pass_size ));
        }

  if( ki.generator != NULL ) {
    ak_ptr_wipe( buf1, sizeof( buf1 ), ki.generator );
    ak_ptr_wipe( buf2, sizeof( buf2 ), ki.generator );
  }

 return buflen1;
}

/* ----------------------------------------------------------------------------------------------- */
 int aktool_remove_file( const tchar *filename )
{
  char ch[8];

  if( ki.confirm ) { /* пользователь хочет подтверждать удаление файла */
    printf(_("Remove the file %s? [y/n]"), filename ); fflush( stdout );
    memset( ch, 0, sizeof( ch ));
    if( fgets( ch, sizeof( ch ) -1, stdin ) == NULL ) return ak_error_null_pointer;

    if( ch[0] == 'y' || ch[0] == 'Y' ) {
      if( remove( filename ) < 0 ) return ak_error_access_file;
      return ak_error_ok;
    }
    return ak_error_cancel_delete_file;
  }
  if( remove( filename ) < 0 ) return ak_error_access_file;
 return ak_error_ok;
}

/* ----------------------------------------------------------------------------------------------- */
/*                                 реализация вывода справки                                       */
/* ----------------------------------------------------------------------------------------------- */
 int aktool_litehelp( void )
{
  printf(_("aktool - crypto utility based on libakrypt library (version: %s)\n\n"),
                                                                          ak_libakrypt_version( ));
  printf(_("try \"aktool --help\" to get more information\n"));

 return EXIT_SUCCESS;
}

/* ----------------------------------------------------------------------------------------------- */
 int aktool_version( void )
{
  printf(_("version: %s [compiled by %s, version: %s, date: %s at %s]\n"),
               ak_libakrypt_version(), LIBAKRYPT_COMPILER_NAME, __VERSION__, __DATE__, __TIME__ );
 return EXIT_SUCCESS;
}

/* ----------------------------------------------------------------------------------------------- */
 int aktool_print_common_options( void )
{
  printf(
   _("\ncommon aktool options:\n"
     "     --audit             set the audit level [ enabled values : 0 (none), 1 (standard), 2 (max) ]\n"
     "     --audit-file        set the output file for errors and libakrypt audit system messages\n"
     "     --confirm           ask for confirmation when deleting files\n"
     "     --dont-use-colors   do not use the highlighting of output data\n"
     "     --help              show this information\n"
     "     --hex-input         read characters from terminal or console as hexademal numbers\n"
     "     --openssl-style     use non-standard variants of some crypto algorithms, as in openssl library\n"
     "     --quiet             not to display any information and use only the return code\n"
     "     --verbose           show the additional information\n\n"));

 return EXIT_SUCCESS;
}

/* ----------------------------------------------------------------------------------------------- */
 int aktool_help( void )
{
  printf(_("aktool - crypto utility based on libakrypt library (version: %s)\n"),
                                                                          ak_libakrypt_version( ));
  printf(_("usage:\n"));
  printf(_("  aktool command [options] [files]\n\n"));
  printf(_("available commands (in short and long forms):\n"));
  printf(_("  a, asn1parse  -  decode and print the ASN.1 data\n"));
  printf(_("  i, icode      -  calculate or check integrity codes\n"));
  printf(_("  k, key        -  key generation and management functions\n"));
  printf(_("  s, show       -  show useful information\n"));
  printf(_("  t, test       -  run performance and correct operation tests\n\n"));
  printf(_("also try:\n"));
  printf(_("  \"aktool command --help\" to get information about command options\n"));
  printf(_("  \"aktool --version\" to get version of aktool\n"));
  printf(_("  \"man aktool\" to get more information about aktool and useful examples\n\n"));
  printf(_("aktool compiled by %s, version: %s (%s at %s)\n"),
                                         LIBAKRYPT_COMPILER_NAME, __VERSION__, __DATE__, __TIME__ );
 return EXIT_SUCCESS;
}

/* ----------------------------------------------------------------------------------------------- */
/*                                                                                       akrypt.c  */
/* ----------------------------------------------------------------------------------------------- */
